src/Storages/ObjectStorage/DataLakes/IcebergMetadata.cpp

```diff
@@ -235,91 +208,230 @@ DataTypePtr getFieldType(const Poco::JSON::Object::Ptr & field, const String & t
+                    "Canßnot add a column with id {} with required values to the table during schema evolution, old schema id is {}, new "
+                    "schema id is {}",
+                    id,
+                    current_old_id,
+                    current_new_id);
+            }
+            ColumnPtr default_type_column = type->createColumnConstWithDefaultValue(0);
+            const auto & constant = dag->addColumn({default_type_column, type, name});
+            outputs.push_back(&constant);
         }
     }
-    else
+    return dag;
+}
+
+std::shared_ptr<const ActionsDAG> IcebergSchemaProcessor::getSchemaTransformationDagByIds(Int32 old_id, Int32 new_id)
+{
+    current_old_id = old_id;
+    current_new_id = new_id;
+    SCOPE_EXIT({
+        current_old_id = -1;
+        current_new_id = -1;
+    });
+    Poco::JSON::Object::Ptr old_schema, new_schema;
+    if (transform_dags_by_ids.contains({old_id, new_id}))
+    {
+        return transform_dags_by_ids.at({old_id, new_id});
+    }
+    try
+    {
+        old_schema = iceberg_table_schemas_by_ids.at(old_id);
+    }
+    catch (std::exception &)
+    {
+        throw Exception(ErrorCodes::BAD_ARGUMENTS, "Schema with schema-id {} is unknown", old_id);
+    }
+    if (old_id == new_id)
+    {
+        return nullptr;
+    }
+    try
+    {
+        new_schema = iceberg_table_schemas_by_ids.at(new_id);
+    }
+    catch (std::exception &)
     {
-        schema = metadata_object->getObject("schema");
-        current_schema_id = schema->getValue<int>("schema-id");
-        /// Field "schemas" is optional for version 1, but after version 2 was introduced,
-        /// in most cases this field is added for new tables in version 1 as well.
-        if (!ignore_schema_evolution && metadata_object->has("schemas") && metadata_object->get("schemas").extract<Poco::JSON::Array::Ptr>()->size() > 1)
-            throw Exception(ErrorCodes::UNSUPPORTED_METHOD,
-                "Cannot read Iceberg table: the table schema has been changed at least 1 time, reading tables with evolved schema is not "
-                "supported. If you want to ignore schema evolution and read all files using latest schema saved on table creation, enable setting "
-                "iceberg_engine_ignore_schema_evolution (Note: enabling this setting can lead to incorrect result)");
+        throw Exception(ErrorCodes::BAD_ARGUMENTS, "Schema with schema-id {} is unknown", new_id);
     }
+    return transform_dags_by_ids[{old_id, new_id}] = getSchemaTransformationDag(old_schema, new_schema);
+}
 
-    NamesAndTypesList names_and_types;
-    auto fields = schema->get("fields").extract<Poco::JSON::Array::Ptr>();
-    for (size_t i = 0; i != fields->size(); ++i)
+void IcebergSchemaProcessor::addIcebergTableSchema(Poco::JSON::Object::Ptr schema_ptr)
+{
+    Int32 schema_id = schema_ptr->getValue<Int32>("schema-id");
+    if (iceberg_table_schemas_by_ids.contains(schema_id))
```

Comment:
Maybe better to use `iceberg_table_schemas_by_ids.emplace(schema_ptr)`. It inserts new value only if there was no such key and returns iterator to the object with a bool flag if it was inserted or not (in this case iterator to existing object is returned)

Reply:
I think that here it is an over-optimization that can reduce the clarity of the code